### PR TITLE
bump kubernetes to 1.9.3

### PR DESCRIPTION
**Pull request by andyxning**
This PR will bump client-go/api/metrics/apimachinery/kubernetes to 1.9.3. 

As node-problem-detector depends on heapster in some pkgs, we need firs to update heapster to support Kubernetes 1.8 first.

Main changes:
* remove source `apiVersion` config as [`Cluster`](https://github.com/kubernetes/client-go/blob/release-5.0/tools/clientcmd/api/types.go#L63-L80) has no `APIVersion` field for now.
* remove usage of `NodeLegacyHostIP` as it has been deprecated.
* deprecate `heapster/metrics/metrics` and use `metrics/pkg/apis/metrics`.
* `k8s.io/client-go/pkg/api` -> `k8s.io/kubernetes/pkg/api`.
* `k8s.io/client-go/pkg/api/v1` -> `k8s.io/api/core/v1`.
* enable RBAC for e2e.

/cc @DirectXMan12 @loburm @fgrzadkowski

### CODE REVIEW

**Comment by loburm** [comment]
@andyxning is there any chance to split this PR into smaller chunks? Or all changes are dependent on each other?

**Comment by andyxning** [comment]
@loburm I have tried to do so. But api&&apimachinery&&metrics&&client-go are referencing each other. It is hard or impossible to only update one.

**Comment by loburm** [comment]
OK, I see could you please resolve conflicts first and then I'll check what we can do.

**Comment by andyxning** [comment]
@loburm Thanks.

Some significant changes are documented in the PR description.

**Comment by loburm** [comment]
@andyxning I think that we also need to change supported_kube_version in Makefile.

**Comment by andyxning** [comment]
@loburm Done.
* Makefile: https://github.com/kubernetes/heapster/pull/1849/files#diff-b67911656ef5d18c4ae36cb6741b7965
* integration/.jenkins.sh: https://github.com/kubernetes/heapster/pull/1849/files#diff-ee3c240603fa28e187c659a556bb782b

**Comment by piosz** [comment]
LGTM in general. Huge thanks @andyxning for doing this super grungy work.

E2e are failing
```
I1018 02:50:00.002] --- FAIL: TestHeapster (591.37s)
I1018 02:50:00.003] 	assertions.go:203: 
                        
	Error Trace:	heapster_api_test.go:1159
I1018 02:50:00.003] 		
	Error:		Received unexpected error "failed to parse heapster controller - no kind \"ReplicationController\" is registered for version \"v1\""
```

The PR also needs to be rebased. Will you find some time to take a look into this?

**Comment by andyxning** [comment]
@piosz Will handle this in about 3 days.

**Comment by piosz** [comment]
@andyxning thanks. I'll release `v1.5.0` then and include this in `v1.5.0`.

**Comment by DirectXMan12** [comment]
for future reference, why does node-problem-detector depend on Heapster?

**Comment by andyxning** [comment]
@piosz Sorry for the delay. But too busy these days. In case you have a schedule for releasing v1.5.0 for heapster, i may have not enough time to make this happen this week. I will make this happen later.

**Comment by piosz** [comment]
@andyxning now worries. As I wrote, we will include those changes in a patch release.

**Comment by andyxning** [comment]
@piosz @loburm @DirectXMan12 Sorry for the late and finally i got it. For now the ci and e2e test are both green. PTAL.

**Comment by andyxning** [comment]
@DirectXMan12 
> for future reference, why does node-problem-detector depend on Heapster?

This is used to add running NPD in standalone mode. FYI: https://github.com/kubernetes/node-problem-detector/pull/49

**Comment by x13n** [comment]
FWIW, when I was trying to bump k8s version in Heapster tests, this is what I found in Heapster logs:

E0115 13:51:27.283122 1 reflector.go:190] k8s.io/heapster/metrics/util/util.go:30: Failed to list *v1.Node: nodes is forbidden: User "system:serviceaccount:heapster-e2e-tests:default" cannot list nodes at the cluster scope
E0115 13:51:27.283299 1 reflector.go:190] k8s.io/heapster/metrics/heapster.go:328: Failed to list *v1.Pod: pods is forbidden: User "system:serviceaccount:heapster-e2e-tests:default" cannot list pods at the cluster scope
E0115 13:51:27.283492 1 reflector.go:190] k8s.io/heapster/metrics/processors/namespace_based_enricher.go:89: Failed to list *v1.Namespace: namespaces is forbidden: User "system:serviceaccount:heapster-e2e-tests:default" cannot list namespaces at the cluster scope

Looks like Heapster running in a separate namespace can no longer list pods or nodes without being explicitly granted those rights.

**Comment by andyxning** [comment]
@x13n  Does this related with RBAC?

**Comment by x13n** [comment]
Yup, that's my guess.

**Comment by andyxning** [comment]
@piosz @DirectXMan12 @loburm @x13n Thank God. I finally make it. PTAL. All CIs are green now.

**Comment by andyxning** [comment]
ping @DirectXMan12 @piosz @loburm @x13n @BenTheElder

File: common/kubernetes/configs.go
Comment (DirectXMan12) [inline common/kubernetes/configs.go]: this should just be `"k8s.io/api/core/v1".SchemeGroupVersion`

File: common/kubernetes/configs.go
Comment (DirectXMan12) [inline common/kubernetes/configs.go]: ditto here

File: metrics/apis/metrics/handlers.go
Comment (DirectXMan12) [inline metrics/apis/metrics/handlers.go]: Nope, we can't do this -- the legacy client still expects the alpha API version on the non-aggregated endpoint.  If we're going to change this, we'd have to just remove it.  We can't really do that unless we also say that the most recent version of Heapster is the last to be supported as a source of metrics for HPA (FWIW, I'm fine with saying that, *but it's something that we need to decide as a team* @kubernetes/heapster-maintainers ).

**Review by DirectXMan12 — CHANGES_REQUESTED** [review]
comments inline.  a couple of nits, and there's some significant discussion that needs to occur around the API handlers (also inline).

Would like to get someone else from @kubernetes/heapster-maintainers to look at the test framework changes, but they're probably fine.

File: common/kubernetes/configs.go
Comment (andyxning) [reply common/kubernetes/configs.go]: It should be `"k8s.io/apimachinery/pkg/apis/meta/v1". SchemeGroupVersion`.

Done

File: common/kubernetes/configs.go
Comment (andyxning) [reply common/kubernetes/configs.go]: Done

File: metrics/apis/metrics/handlers.go
Comment (andyxning) [reply metrics/apis/metrics/handlers.go]: > the legacy client still expects the alpha API version on the non-aggregated endpoint.

The client is HPA?

/cc @piosz @loburm

**Comment by andyxning** [comment]
Ping @kubernetes/heapster-maintainers . PTAL

File: metrics/apis/metrics/handlers.go
Comment (x13n) [reply metrics/apis/metrics/handlers.go]: /cc @MaciekPytel @kawych to address the HPA question.

Comment (kawych) [reply metrics/apis/metrics/handlers.go]: The legacy client is used by HPA and kubectl top. The latter case was removed ~1 week ago in #56206. I agree with @DirectXMan12 that we shouldn't change the API version, instead we should remove this endpoint entirely from Heapster at some point (I'd say it's definetely too early for that).

Comment (piosz) [reply metrics/apis/metrics/handlers.go]: +1 for not bumping the version, as kubectl top relied on this until very recently.

File: metrics/apis/metrics/handlers.go
Comment (piosz) [inline metrics/apis/metrics/handlers.go]: same here - see the comment above

File: metrics/cmd/heapster-apiserver/app/server.go
Comment (piosz) [inline metrics/cmd/heapster-apiserver/app/server.go]: Consider removing support for aggregated apiserver as this is currently implemented by metrics-server.

@DirectXMan12 WDYT?

**Review by piosz — DISMISSED** [review]
Two comments:
1. please revert change bumping metrics api from v1alpha1 to v1beta1
2. consider removing support for aggregated apiserver in Heapster

Otherwise LGTM

**Comment by andyxning** [comment]
@piosz @DirectXMan12 
> Two comments:

> please revert change bumping metrics api from v1alpha1 to v1beta1
consider removing support for aggregated apiserver in Heapster
Otherwise LGTM

* `please revert change bumping metrics api from v1alpha1 to v1beta1`. Done with using `k8s.io/metrics/pkg/apis/metrics/v1alpha1`.
* `consider removing support for aggregated apiserver in Heapster`. Pending as i am not familiar with this. I think we can do this in another PR. :)

Ping @jingxu97 for reviewing this PR. :)

**Comment by andyxning** [comment]
The integration failed. Can anyone see it and give me some advise? Too complicated. :(

**Comment by jingxu97** [comment]
cc @piosz @kawych

File: deploy/kube-config/standalone-test/heapster-service-account.yaml
Comment (kawych) [inline deploy/kube-config/standalone-test/heapster-service-account.yaml]: Please use a namespace "heapster-e2e-tests" here

File: integration/heapster_api_test.go
Comment (kawych) [inline integration/heapster_api_test.go]: Please use different rbac file, for your testing service account. The current one binds cluster role system:heapster to the service account from kube-system namespace.

File: integration/heapster_api_test.go
Comment (kawych) [inline integration/heapster_api_test.go]: You will have to call setupHeapsterServiceAccount and setupHeapsterRBAC after the test namespace is created.

**Comment by kawych** [comment]
@jingxu97 @andyxning 
I've debugged the integration test for a bit, it's an issue in your deployment files that causes heapster pod not to be created. I've left some comments, after applying these changes, heapster pod is created and tests are running, so you should be fine.

File: deploy/kube-config/standalone-test/heapster-service-account.yaml
Comment (andyxning) [reply deploy/kube-config/standalone-test/heapster-service-account.yaml]: Will do.

File: integration/heapster_api_test.go
Comment (andyxning) [reply integration/heapster_api_test.go]: Added a new heapster-rbac.yaml file to standalone-test dir.

File: integration/heapster_api_test.go
Comment (andyxning) [reply integration/heapster_api_test.go]: Will do.

**Comment by andyxning** [comment]
@kawych Thanks for the diagnose and suggestions. After all, there are still some problems. Could you please help to take a look at this?

File: deploy/kube-config/standalone-test/heapster-controller.yaml
Comment (kawych) [inline deploy/kube-config/standalone-test/heapster-controller.yaml]: We should probably bump the image version to v1.5.1

**Comment by andyxning** [comment]
Finally ping @loburm @piosz @DirectXMan12 @x13n @jingxu97 

Sorry for the quite long time delay but with the help from @kawych we finally make all the ci green. PTAL.

**Comment by x13n** [comment]
Yay! Looks good to me, deferring to others who previously had comments to apply lgtm label.

**Comment by kawych** [comment]
@piosz @DirectXMan12 - Please take a look and re-apply LGTM label if it looks fine
@andyxning - Note #1980 Consider bumping it one more version - to 1.9.4. Alternatively we can get this merged and bump it again later.

**Comment by x13n** [comment]
I'm against bumping the version again - there is already too much happening in this PR.

**Comment by andyxning** [comment]
@x13n @kawych I am also suggest we first review and merge this PR.

/cc @piosz @DirectXMan12 PTAL ASAP. :)

**Comment by andyxning** [comment]
@loburm @kawych @x13n Can anyone of you give piosz a DM since you are both Googlers. :)

File: integration/framework.go
Comment (loburm) [inline integration/framework.go]: Could you add comments here and below?

File: metrics/cmd/heapster-apiserver/app/server.go
Comment (loburm) [reply metrics/cmd/heapster-apiserver/app/server.go]: I think it's OK to keep it in this PR. Let's not try to overcomplicate it.

**Review by loburm — COMMENTED** [review]
Piotr asked me to finish this review. Please add last comments and then lgtm.

File: deploy/kube-config/standalone-test/heapster-controller.yaml
Comment (andyxning) [reply deploy/kube-config/standalone-test/heapster-controller.yaml]: Done.

File: integration/framework.go
Comment (andyxning) [reply integration/framework.go]: Done. PTAL.

**Comment by jingxu97** [comment]
@loburm @piosz @DirectXMan12 Could you please take a look and see whether this PR is ready to be merged? Thanks a lot!

**Comment by jiayingz** [comment]
Co-ping. Given that there are other PRs queued after this one, can we speed up the review on this PR?

**Pull request merged**
